### PR TITLE
Rework derivative algorithm completely

### LINKED ISSUE #13570: DerivativeParsedMaterial start-up performance issues
## Bug Description
 I have identified two bottle necks in the startup (differentiation phase) if DerivativeParsedMaterials. One takes 26.6% of the entire derivation time, and one takes 68.3%. Both involve an unnecessary amount of string operations that I may be able to trim
down considerably.

## Steps to Reproduce
Run a simulation with DerivativeParsedMaterials that use lots of material properties that are provided by DerivativeParsedMaterials themselves (chain rule).

## Impact
Performance improvement.

### PR DESCRIPTION
**Pull request by dschwen**
Using civet to test this for me...

I still need to run this through a profiler, but initial results are promising. @jbair34's input that didn't finish taking derivatives after two days finishes setup in 90 seconds now.

Closes #13570

### CODE REVIEW

**Comment by dschwen** [comment]
Ok, This is ready to go. The new algorithm is much clearer than the convoluted (and ill designed) deque algorithm) and it has a much lower computational complexity (lower O(N) exponent).

Running this through a profiler shows me that the main cost is now down in FParser. I will tackle that in a libmesh PR later.

**Comment by dschwen** [comment]
Hold up, I'm making one last change...

**Comment by dschwen** [comment]
With the last commit @jbair34's input that didn't finish taking derivatives after two days finishes setup in 10 seconds now!

**Comment by dschwen** [comment]
@permcody , I suspect you are super busy. If you can suggest an alternate reviewer that'd be great.

File: framework/include/materials/DerivativeParsedMaterialHelper.h [hunk 1]
```diff
@@ -48,24 +56,26 @@ class DerivativeParsedMaterialHelper : public ParsedMaterialHelper
...
+  std::vector<VariableName> _darg_names;
 };
 
-struct DerivativeParsedMaterialHelper::Derivative
+struct DerivativeParsedMaterialHelper::MaterialPropertyDerivativeRule
 {
-  MaterialProperty<Real> * first;
-  ADFunctionPtr second;
-  std::vector<VariableName> darg_names;
+  MaterialPropertyDerivativeRule(std::string p, std::string v, std::string c)
```
Comment (permcody) [inline framework/include/materials/DerivativeParsedMaterialHelper.h]: You could avoid this copy

File: framework/src/materials/DerivativeParsedMaterialHelper.C [hunk 2]
```diff
@@ -54,21 +54,95 @@ DerivativeParsedMaterialHelper::functionsPostParse()
...
+
+  // variable we are deriving w.r.t.
+  auto derivative_var = _variable_names[var];
+
+  // current derivative starts off of the parent function
+  Derivative current;
+  current._darg_names = parent_derivative._darg_names;
+  // the moose variable name goes into the derivative property name
+  current._darg_names.push_back(_arg_names[var]);
+  current._F = ADFunctionPtr(new ADFunction(*parent_derivative._F));
```
Comment (permcody) [inline framework/src/materials/DerivativeParsedMaterialHelper.C]: `std::make_shared` here and throughout?

**Review by permcody — APPROVED** [review]
This looks good - two minor comments but nothing worth holding this up for. Feel free to merge or fixup as you see fit.

**Comment by dschwen** [comment]
Thanks, I'll go ahead and merge this, but I will get back to the copy and make::shared issue (I'll open a ticket for those). Ah, I was just racking my brain about how to use move semantics here and you probably just meant using `const std::string &` as argument types...

**Pull request merged**